### PR TITLE
Adding Notebooks for Fossils case study

### PR DESCRIPTION
**Pull request by nareshsusarla**
## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.md and COPYRIGHT.md file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by ksbeattie** [comment]
This needs PR #99

**Comment by lbianchi-lbl** [comment]
@nareshsusarla @radhakrishnatg I've created a [Wiki page](https://github.com/gmlc-dispatches/dispatches/wiki/How-to:-use-RTS-GMLC-data-in-DISPATCHES-examples) with instructions on using the RTS-GMLC data. When you have a moment, please try it locally and see if it works within the notebook(s) where the data is required.

**Comment by radhakrishnatg** [comment]
@lbianchi-lbl Here is the complete error message:
```
NameError                                 Traceback (most recent call last)
<ipython-input-7-f4a3ab150562> in <module>
     32 
     33 # Run Prescient
---> 34 Prescient().simulate(**options)

c:\users\radhakrishna\desktop\prescient\prescient\simulator\prescient.py in simulate(self, **options)
     73             self.config.set_value(options)
     74 
---> 75         return self._simulate(self.config)
     76 
     77     def _simulate(self, options: PrescientConfig):

c:\users\radhakrishna\desktop\prescient\prescient\simulator\prescient.py in _simulate(self, options)
     80                                 "If you wish to simulate again create a new Prescient object.")
     81         self.simulate_called = True
---> 82         return super().simulate(options)
     83 
     84 def main(args=None):

c:\users\radhakrishna\desktop\prescient\prescient\simulator\simulator.py in simulate(self, options)
     83         self.callback_manager.invoke_options_preview_callbacks(options)
     84 
---> 85         engine.initialize(options)
...
--> 157         raise ValueError(f'RTS-GMLC directory "{rts_gmlc_dir}" does not contain expected CSV files.')
    158 
    159     # Read metadata about the data

NameError: name 'rts_gmlc_dir' is not defined
```
Should we ask Ben if Prescient needs any other files apart from those in that folder?

**Comment by lbianchi-lbl** [comment]
@radhakrishnatg Thanks, but I still can't understand where the error is raised actually is.

I see something called `self.callback_manager` makes me think that there could be some dynamically loaded code, but searching for the error message in both this repository and grid-parity-exchange/Prescient is not turning up any match.  @bknueven, would you happen to have more information about this error message and/or how I can take a look at the code that produces it?

**Comment by lbianchi-lbl** [comment]
OK, I think I find the reason for the error - I forgot to include one file in the `dispatches-sample-data` package (luckily, it was exactly the one that's being checked by Egret [here](https://github.com/grid-parity-exchange/Egret/blob/12e37578f5337c01961a97c89c26121f321e29b5/egret/parsers/rts_gmlc/parser.py#L154).

I'll let you know as soon as a patched version will be available.

**Comment by lbianchi-lbl** [comment]
@radhakrishnatg OK, I have a test package up that seems to have solved at least this particular issue. Can you try to run these commands and verify that they work without errors?

```py
pip install -i https://test.pypi.org/simple/ --extra-index-url https://pypi.org/simple "dispatches-sample-data==22.6.23.1"
python -c "from dispatches_sample_data import rts_gmlc as data; from egret.parsers.rts_gmlc.parser import _read_metadata as read; print(read(data.path))"
```

**Comment by bknueven** [comment]
@lbianchi-lbl see this commit for @radhakrishnatg's error: https://github.com/bknueven/RTS-GMLC/commit/d93b196f664a1dbff72bc5a255c9d8c54832a268.

Too many of us use macOS and its insane (default) case-insensitive file system.

**Comment by lbianchi-lbl** [comment]
@bknueven and @radhakrishnatg , I've just uploaded a new version of the package (22.06.23.3), that bundles all necessary CSV files from the RTS-GMLC repository. You can install by running:

```py
pip install -i https://test.pypi.org/simple/ dispatches-sample-data
python -c "from dispatches_sample_data import __version__ as v; print(v)"
```

@radhakrishnatg , could you try running the notebook again with 22.06.23.3? The case-insensitivity issue sounds plausible (if so, I'll have to create another version after pulling the fixed data files from @bknueven's fork) but I'm not sure if this particular error isn't due to files that were actually missing from 22.06.23.1.

**Comment by radhakrishnatg** [comment]
@lbianchi-lbl I got the following error message:
```
(base) C:\Users\Radhakrishna>python -c "from dispatches_sample_data import __version__ as v; print(v)"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\Radhakrishna\Anaconda3\lib\site-packages\dispatches_sample_data\__init__.py", line 3, in <module>
    from importlib import metadata
ImportError: cannot import name 'metadata' from 'importlib' (C:\Users\Radhakrishna\Anaconda3\lib\importlib\__init__.py)
```

**Comment by lbianchi-lbl** [comment]
> @lbianchi-lbl I got the following error message:
> 
> ```
> (base) C:\Users\Radhakrishna>python -c "from dispatches_sample_data import __version__ as v; print(v)"
> Traceback (most recent call last):
>   File "<string>", line 1, in <module>
>   File "C:\Users\Radhakrishna\Anaconda3\lib\site-packages\dispatches_sample_data\__init__.py", line 3, in <module>
>     from importlib import metadata
> ImportError: cannot import name 'metadata' from 'importlib' (C:\Users\Radhakrishna\Anaconda3\lib\importlib\__init__.py)
> ```

@radhakrishnatg sorry about that, I think I might have used the wrong exception handling (`ModuleNotFoundError` when it should be the more general `ImportError`). I assume you're using Python 3.7? I'll update this comment as soon as I release a new version with this patch.

**Comment by radhakrishnatg** [comment]
@lbianchi-lbl Yes, I'm using Python 3.7. Sorry, I need to update it to 3.10.
Now, the commands you mentioned run without any errors. Unfortunately, that is not the case with the notebook.

You mentioned that "the directory structure in `dispatches_sample_data.rts_gmlc` has changed, so you'll probably have to change the variable passed to the Prescient options dict to either `dispatches_sample_data.rts_gmlc / "SourceData"` or the convenience alias `dispatches_sample_data.rts_gmlc.source_data_path`." 

But it gives me the following error: `ValueError: RTS-GMLC directory "C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\SourceData" does not exist`

I also tried running the notebook with the old command `rts_gmlc_data_dir = str(rts_gmlc.source_data_path)`, and it gave me the same error before.
```
-----------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-10-f4a3ab150562> in <module>
     32 
     33 # Run Prescient
---> 34 Prescient().simulate(**options)

c:\users\radhakrishna\desktop\prescient\prescient\simulator\prescient.py in simulate(self, **options)
     73             self.config.set_value(options)
     74 
---> 75         return self._simulate(self.config)
     76 
     77     def _simulate(self, options: PrescientConfig):

c:\users\radhakrishna\desktop\prescient\prescient\simulator\prescient.py in _simulate(self, options)
     80                                 "If you wish to simulate again create a new Prescient object.")
     81         self.simulate_called = True
---> 82         return super().simulate(options)
     83 
     84 def main(args=None):

c:\users\radhakrishna\desktop\prescient\prescient\simulator\simulator.py in simulate(self, options)
     83         self.callback_manager.invoke_options_preview_callbacks(options)
     84 
---> 85         engine.initialize(options)
...
--> 647                 newline="",
    648             )
    649         else:

FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\Radhakrishna\\Anaconda3\\Lib\\site-packages\\dispatches_sample_data\\timeseries_data_files\\HYDRO\\DAY_AHEAD_hydro.csv'
```

**Comment by lbianchi-lbl** [comment]
Thanks for checking @radhakrishnatg 

> @lbianchi-lbl Yes, I'm using Python 3.7. Sorry, I need to update it to 3.10.

No worries, the newest version of `dispatches-sample-data` should work on any of the Python versions that are supported by DISPATCHES.

> Now, the commands you mentioned run without any errors. Unfortunately, that is not the case with the notebook.
> You mentioned that "the directory structure in `dispatches_sample_data.rts_gmlc` has changed, so you'll probably have to change the variable passed to the Prescient options dict to either `dispatches_sample_data.rts_gmlc / "SourceData"` or the convenience alias `dispatches_sample_data.rts_gmlc.source_data_path`."
> 
> But it gives me the following error: `ValueError: RTS-GMLC directory "C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\SourceData" does not exist`

Yes, sorry, I think I made a mistake in my instructions in that comment. `rts_gmlc.path / "SourceData"` should work (I think the `.path` before `/ "SourceData"` was missing). In any case, it's probably better to use `rts_gmlc.source_data_path` (or `str(rts_gmlc.source_data_path)` if you need a string instead of a `Path` object).

> I also tried running the notebook with the old command `rts_gmlc_data_dir = str(rts_gmlc.source_data_path)`, and it gave me the same error before.
> 
> ```
> -----------------------------------------------------------
> FileNotFoundError                         Traceback (most recent call last)
> <ipython-input-10-f4a3ab150562> in <module>
>      32 
>      33 # Run Prescient
> ---> 34 Prescient().simulate(**options)
> 
> c:\users\radhakrishna\desktop\prescient\prescient\simulator\prescient.py in simulate(self, **options)
>      73             self.config.set_value(options)
>      74 
> ---> 75         return self._simulate(self.config)
>      76 
>      77     def _simulate(self, options: PrescientConfig):
> 
> c:\users\radhakrishna\desktop\prescient\prescient\simulator\prescient.py in _simulate(self, options)
>      80                                 "If you wish to simulate again create a new Prescient object.")
>      81         self.simulate_called = True
> ---> 82         return super().simulate(options)
>      83 
>      84 def main(args=None):
> 
> c:\users\radhakrishna\desktop\prescient\prescient\simulator\simulator.py in simulate(self, options)
>      83         self.callback_manager.invoke_options_preview_callbacks(options)
>      84 
> ---> 85         engine.initialize(options)
> ...
> --> 647                 newline="",
>     648             )
>     649         else:
> 
> FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\Radhakrishna\\Anaconda3\\Lib\\site-packages\\dispatches_sample_data\\timeseries_data_files\\HYDRO\\DAY_AHEAD_hydro.csv'
> ```

OK, so this might be related to the error @bknueven mentioned, but to confirm this, we have to make sure that the files actually exist (with incompatible names). Can you run (or re-run) the following checks, also posting the output?

1. Run `where python`
1. Run `python -c "import sys; print(sys.executable)"`
1. Run `python -c "from dispatches_sample_data import __version__ as v; print(v)"` (to make sure that the installed version is 22.06.23.4)
1. `python -c "from dispatches_sample_data import rts_gmlc as data; [print(p) for p in sorted(data.path.rglob('*.csv'))]"` (this will list the CSV file names as they appear to the Python code)
1. Using the graphical Windows file explorer, navigate to e.g. `C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\timeseries_data_files`, and take a screenshot (this is to help investigate if the case-sensitiveness of the names is causing problems)

Finally, I've also noticed that the path names suggest that the packages are being installed and run in the Conda `base` environment. We generally recommend to create and use a separate environment (e.g. `conda create -n dispatches-dev --yes python && conda activate dispatches-dev && pip install -r requirements-dev.txt`). My suggestion is for the time being to keep using the environment you're currently working in, and if it turns out that this is causing issue I can help set up a separate environment later.

**Comment by radhakrishnatg** [comment]
Thank you @lbianchi-lbl Here is the output for steps 1 - 4:
```
(base) C:\Users\Radhakrishna>where python
C:\Users\Radhakrishna\Anaconda3\python.exe
C:\Users\Radhakrishna\AppData\Local\Microsoft\WindowsApps\python.exe

(base) C:\Users\Radhakrishna>python -c "import sys; print(sys.executable)"
C:\Users\Radhakrishna\Anaconda3\python.exe

(base) C:\Users\Radhakrishna>python -c "from dispatches_sample_data import __version__ as v; print(v)"
22.6.23.4

(base) C:\Users\Radhakrishna>python -c "from dispatches_sample_data import rts_gmlc as data; [print(p) for p in sorted(data.path.rglob('*.csv'))]"
C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\branch.csv
C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\bus.csv
C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\dc_branch.csv
C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\gen.csv
C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\reserves.csv
C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\simulation_objects.csv
C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\storage.csv
C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\timeseries_pointers.csv

(base) C:\Users\Radhakrishna>
```
I cannot find the `timeseries_data_files` folder in the `dispatches_sample_data` folder. 
Thank you for the help @lbianchi-lbl ! I didn't know about conda environments in the beginning, so I was using `base` for everything. I will start using a new laptop in a couple of days. I set up a new environment for dispatches on the new laptop. I have a couple of questions regarding that, so I will contact you on slack later.

**Comment by lbianchi-lbl** [comment]
Thanks @radhakrishnatg for running the checks. It looks like there might be some inconsistency: from (2), it seems the right version of `dispatches-sample-data` is installed; however, the output of (4) shows a file structure that belongs to an older version.

Could you run the next batch of checks, and post their outputs?

1. `pip list`
2. `pip uninstall dispatches-sample-data --yes`
3. `pip list`
4. `pip install -i https://test.pypi.org/simple/ dispatches-sample-data==22.06.23.4`
5. `pip list`
6. `python -c "from dispatches_sample_data import rts_gmlc as data; [print(p) for p in sorted(data.path.rglob('*.csv'))]"`

**Comment by lbianchi-lbl** [comment]
@nareshsusarla I've looked at the tests that are currently failing:

- There seems to be an undefined `solver` objects here:
  https://github.com/gmlc-dispatches/dispatches/blob/a8cd6b12cfb448bec5330509144fed5bae255e90/dispatches/models/fossil_case/ultra_supercritical_plant/storage/integrated_storage_with_ultrasupercritical_power_plant.py#L942
- The other failure is due to a JSON file not being found:
  https://github.com/gmlc-dispatches/dispatches/blob/a8cd6b12cfb448bec5330509144fed5bae255e90/dispatches/models/fossil_case/ultra_supercritical_plant/storage/integrated_storage_with_ultrasupercritical_power_plant.py#L1110

I'll look into this to see if the missing JSON file is due to a path and/or packaging issue.

**Comment by nareshsusarla** [comment]
> @nareshsusarla I've looked at the tests that are currently failing:
> 
> * There seems to be an undefined `solver` objects here:
>   https://github.com/gmlc-dispatches/dispatches/blob/a8cd6b12cfb448bec5330509144fed5bae255e90/dispatches/models/fossil_case/ultra_supercritical_plant/storage/integrated_storage_with_ultrasupercritical_power_plant.py#L942
> * The other failure is due to a JSON file not being found:
>   https://github.com/gmlc-dispatches/dispatches/blob/a8cd6b12cfb448bec5330509144fed5bae255e90/dispatches/models/fossil_case/ultra_supercritical_plant/storage/integrated_storage_with_ultrasupercritical_power_plant.py#L1110
> 
> I'll look into this to see if the missing JSON file is due to a path and/or packaging issue.

@lbianchi-lbl Interestingly, the file causing the issue is not even part of this PR. It is merged into main. Also, I merged latest main into this branch and yet I see the error here.

**Comment by radhakrishnatg** [comment]
@lbianchi-lbl Here is the final output. The contents of the directory didn't seem to change;

```
(base) C:\Users\Radhakrishna>python -c "from dispatches_sample_data import rts_gmlc as data; [print(p) for p in sorted(data.path.rglob('*.csv'))]"
C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\branch.csv
C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\bus.csv
C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\dc_branch.csv
C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\gen.csv
C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\reserves.csv
C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\simulation_objects.csv
C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\storage.csv
C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\timeseries_pointers.csv
```

**Comment by lbianchi-lbl** [comment]
> @lbianchi-lbl Here is the final output. The contents of the directory didn't seem to change;
> 
> ```
> (base) C:\Users\Radhakrishna>python -c "from dispatches_sample_data import rts_gmlc as data; [print(p) for p in sorted(data.path.rglob('*.csv'))]"
> C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\branch.csv
> C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\bus.csv
> C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\dc_branch.csv
> C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\gen.csv
> C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\reserves.csv
> C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\simulation_objects.csv
> C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\storage.csv
> C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\rts_gmlc\timeseries_pointers.csv
> ```

@radhakrishnatg Do you happen to have the output from the intermediate commands? I'd need that to help figure out why that directory seems to be stuck at the old structure.

**Comment by radhakrishnatg** [comment]
[output.txt](https://github.com/gmlc-dispatches/dispatches/files/8980556/output.txt)
@lbianchi-lbl Here is the entire output.

**Comment by lbianchi-lbl** [comment]
@radhakrishnatg There is an unexpected error that could explain why the directory (or at least some of the files) are not being deleted:

```txt
(base) C:\Users\Radhakrishna>pip uninstall dispatches-sample-data --yes
Found existing installation: dispatches-sample-data 22.6.23.4
Uninstalling dispatches-sample-data-22.6.23.4:
ERROR: Exception:
Traceback (most recent call last):
  File "C:\Users\Radhakrishna\Anaconda3\lib\shutil.py", line 566, in move
    os.rename(src, real_dst)
PermissionError: [WinError 5] Access is denied: 'c:\\users\\radhakrishna\\anaconda3\\lib\\site-packages\\dispatches_sample_data\\' -> 'c:\\users\\radhakrishna\\anaconda3\\lib\\site-packages\\~ispatches_sample_data'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Radhakrishna\Anaconda3\lib\site-packages\pip\_internal\cli\base_command.py", line 189, in _main
    status = self.run(options, args)
  File "C:\Users\Radhakrishna\Anaconda3\lib\site-packages\pip\_internal\commands\uninstall.py", line 88, in run
    auto_confirm=options.yes, verbose=self.verbosity > 0,
  File "C:\Users\Radhakrishna\Anaconda3\lib\site-packages\pip\_internal\req\req_install.py", line 658, in uninstall
    uninstalled_pathset.remove(auto_confirm, verbose)
  File "C:\Users\Radhakrishna\Anaconda3\lib\site-packages\pip\_internal\req\req_uninstall.py", line 400, in remove
    moved.stash(path)
  File "C:\Users\Radhakrishna\Anaconda3\lib\site-packages\pip\_internal\req\req_uninstall.py", line 289, in stash
    renames(path, new_path)
  File "C:\Users\Radhakrishna\Anaconda3\lib\site-packages\pip\_internal\utils\misc.py", line 347, in renames
    shutil.move(old, new)
  File "C:\Users\Radhakrishna\Anaconda3\lib\shutil.py", line 578, in move
    rmtree(src)
  File "C:\Users\Radhakrishna\Anaconda3\lib\shutil.py", line 516, in rmtree
    return _rmtree_unsafe(path, onerror)
  File "C:\Users\Radhakrishna\Anaconda3\lib\shutil.py", line 395, in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
  File "C:\Users\Radhakrishna\Anaconda3\lib\shutil.py", line 400, in _rmtree_unsafe
    onerror(os.unlink, fullname, sys.exc_info())
  File "C:\Users\Radhakrishna\Anaconda3\lib\shutil.py", line 398, in _rmtree_unsafe
    os.unlink(fullname)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'c:\\users\\radhakrishna\\anaconda3\\lib\\site-packages\\dispatches_sample_data\\rts_gmlc\\timeseries_pointers.csv'
```

Is it possible that the file was open in another program (possibly the one used to open the file to create the screenshot above)?

If so, you can try to:

- Manually delete the `C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\dispatches_sample_data\` folder using e.g. Windows Explorer (and make sure it's really deleted, i.e. navigate to `C:\Users\Radhakrishna\Anaconda3\Lib\site-packages\` and delete the `dispatches_sample_data` subfolder with all its contents
- Then, run steps 1 through 6 again (step 2 might throw an error again, but it's important that it is run and we are able to check the output)

**Comment by radhakrishnatg** [comment]
@lbianchi-lbl Yes, one of the files was open at that time, so it printed the error message. I closed the file and re-ran the commands at that time.

I also did what you suggested. Manually delete the files and repeat the installation process. Here is the output. Again, there is no change in the content of the directory.
[output.txt](https://github.com/gmlc-dispatches/dispatches/files/8980670/output.txt)

**Comment by lbianchi-lbl** [comment]
> @lbianchi-lbl Yes, one of the files was open at that time, so it printed the error message. I closed the file and re-ran the commands at that time.
> 
> I also did what you suggested. Manually delete the files and repeat the installation process. Here is the output. Again, there is no change in the content of the directory. [output.txt](https://github.com/gmlc-dispatches/dispatches/files/8980670/output.txt)

@radhakrishnatg I see, thanks for running the checks for the n-th time. I'm not sure I have other things to suggest to try to investigate this issue at this point, and I wouldn't want to take up any more of your time without first having a clearer idea of what's going on.

My suggestion for the next steps would be:

- I'll try to fix the issue on my side, by trying to run the notebooks (or at least the double-loop workflow) in a Windows environment
- You can go ahead and continue working (and reviewing) the notebooks in the meantime without worrying about using the source of the RTS-GMLC dataset. As a stop-gap solution, I suggest to modify the notebooks with something like this:
  
image

This clearly would still not work out of the box for users (and it doesn't help with being able to run the notebooks programmatically for automatic testing), but at least it's more helpful for users to enable them to test the notebooks locally.

(just to be clear: I still think and plan that we'll manage to fix the remaining issues with the `dispatches-sample-data` package and we'll be able to access the data without hardcoded paths in time for the release).

The Markdown text for the screenshot above is:

```md
**IMPORTANT** Before proceeding with the rest of the notebook:

1. Download the RTS-GMLC dataset from [this repository](https://github.com/GridMod/RTS-GMLC) (or directly as a zipfile [here](https://github.com/GridMod/RTS-GMLC/archive/refs/heads/master.zip)) to your local machine
2. Update the value of the `rst_gmlc_data_dir` variable in the cell below to the appropriate location of the `RTS_Data/SourceData` subfolder on your local machine
```

(This is probably easier/safer than trying to mess with me pushing these changes to all the notebooks in all open PRs).

**Comment by radhakrishnatg** [comment]
Thank you so much @lbianchi-lbl for your help! I'm okay with the approach you suggested, so I will add the comment to the notebook, push the remaining changes and get it reviewed.

**Comment by nareshsusarla** [comment]
@lbianchi-lbl I get the following error when opening the notebook double_loop_integrated_ultrasupercritical_power_plant.ipynb :
`Unreadable Notebook: C:\gmlc\source_code\dispatches\dispatches\models\fossil_case\ultra_supercritical_plant\storage\double_loop_integrated_ultrasupercritical_power_plant.ipynb NotJSONError('Notebook does not appear to be JSON: \'{\\n "cells": [\\n {\\n "cell_type": "m...')`
SO, it will be great if you could help/advice me on how to fix this. Thanks.

**Comment by lbianchi-lbl** [comment]
> @lbianchi-lbl I get the following error when opening the notebook `double_loop_integrated_ultrasupercritical_power_plant.ipynb` : `Unreadable Notebook: C:\gmlc\source_code\dispatches\dispatches\models\fossil_case\ultra_supercritical_plant\storage\double_loop_integrated_ultrasupercritical_power_plant.ipynb NotJSONError('Notebook does not appear to be JSON: \'{\\n "cells": [\\n {\\n "cell_type": "m...')` So, it will be great if you could help/advice me on how to fix this. Thanks.

@nareshsusarla There is a left-over Git merge conflict block at the very end of the file. You should be able to open it with a text editor (you might have to choose e.g. Notepad, or temporarily change the file extension from `.ipynb` to `.txt`,  if your IDE complains about invalid file format) and remove the conflict markup. I think the block from `main` is just an empty cell, so you should be able to just remove it altogether.

```
<<<<<<< HEAD
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e1ee06d",
   "metadata": {},
   "outputs": [],
   "source": []
=======
>>>>>>> 0965fbb7d582546b7621f5e6c75f398aad4ce17e
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
```

**Comment by radhakrishnatg** [comment]
@lbianchi-lbl I installed the new `dispatches-sample-data` package and ran the notebook. I got this error this time:
```
---------------------------------------------------------------------------
PermissionError                           Traceback (most recent call last)
<ipython-input-8-3078a83579e4> in <module>
     32 
     33 # Run Prescient
---> 34 Prescient().simulate(**options)

c:\users\radhakrishna\desktop\prescient\prescient\simulator\prescient.py in simulate(self, **options)
     73             self.config.set_value(options)
     74 
---> 75         return self._simulate(self.config)
     76 
     77     def _simulate(self, options: PrescientConfig):

c:\users\radhakrishna\desktop\prescient\prescient\simulator\prescient.py in _simulate(self, options)
     80                                 "If you wish to simulate again create a new Prescient object.")
     81         self.simulate_called = True
---> 82         return super().simulate(options)
     83 
     84 def main(args=None):

c:\users\radhakrishna\desktop\prescient\prescient\simulator\simulator.py in simulate(self, options)
     88         oracle_manager.initialize(engine, data_manager, options)
     89         stats_manager.initialize(options)
---> 90         reporting_manager.initialize(options, stats_manager)
...
--> 398                 os.unlink(fullname)
    399             except OSError:
    400                 onerror(os.unlink, fullname, sys.exc_info())

PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'c:\\Users\\Radhakrishna\\Desktop\\DISPATCHES\\dispatches\\models\\nuclear_case\\flowsheets\\bidding_plugin_test_multiperiod_nulcear\\bus_detail.csv'
```

**Comment by lbianchi-lbl** [comment]
@radhakrishnatg Thanks for checking. From a first look this seems to be caused by the fact that Prescient's `ReportingManager` objects delete their `output_directory` upon initialization. I can't see exactly what's happening (i.e. how the various part of the `Simulator` used in the notebooks are constructed) until the notebook is made valid again as per https://github.com/gmlc-dispatches/dispatches/pull/119#issuecomment-1167571874.

**Comment by lbianchi-lbl** [comment]
@nareshsusarla I see that there are some failures that seem to be due to data access issues. I'll look into them and find a fix first thing in the morning.

**Comment by nareshsusarla** [comment]
@lbianchi-lbl I used the trick you suggested for json file, but I am getting the following error.
```
# create a pathlib.Path object pointing to the directory
>       dir_path = Path(storage.__path__[0]).resolve()
E       TypeError: '_NamespacePath' object is not subscriptable
```

**Comment by nareshsusarla** [comment]
> ```gfm
> https://github.com/GridMod/RTS-GMLC
> ```

Thanks.

File: dispatches/models/fossil_case/ultra_supercritical_plant/storage/multiperiod_integrated_storage_usc.py [hunk 1]
```diff
@@ -42,8 +43,16 @@ def create_usc_model(pmin, pmax):
     max_power = 436  # in MW
     min_power = int(0.65 * max_power)  # 283 in MW
 
+    # create a pathlib.Path object pointing to the directory
+    dir_path = Path(storage.__path__[0]).resolve()
```
Comment (lbianchi-lbl) [inline dispatches/models/fossil_case/ultra_supercritical_plant/storage/multiperiod_integrated_storage_usc.py:47]:
```suggestion
    dir_path = Path(storage.__path__._path[0]).resolve()
```

File: dispatches/models/fossil_case/ultra_supercritical_plant/storage/multiperiod_double_loop_usc.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+        blk = b
+        if not blk.is_constructed():
+            blk.construct()
+
+        multiperiod_usc = create_multiperiod_usc_model(
+            n_time_points=self.horizon,
+            pmin=self.model_data.p_min, pmax=self.model_data.p_max
+        )
+        blk.usc_mp = multiperiod_usc
+        blk.rankine_model = multiperiod_usc.pyomo_model
```
Comment (radhakrishnatg) [inline dispatches/models/fossil_case/ultra_supercritical_plant/storage/multiperiod_double_loop_usc.py:71]: Please change the variable to usc

Comment (nareshsusarla) [reply dispatches/models/fossil_case/ultra_supercritical_plant/storage/multiperiod_double_loop_usc.py]: The variable is not needed so it is now removed

**Review by radhakrishnatg — APPROVED** [review]
Looks good.

**Review by jghouse88 — APPROVED** [review]

**Pull request merged**